### DISCUSSION TITLE
best strategy for adding push notifications?

### DISCUSSION THREAD

**Discussion by anngbaum** [supabase/supabase#13930] [General]
We have a native mobile app (live on iOS and Android) and we use push notifications via [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging/concept-options) (FCM)

We have generally implemented this with a Google Cloud Function (GCF) written in JS (built before SB edge functions existed – but open to migrating!) that applies a little bit of business logic, and then sends a request to Firebase using the `firebase-admin` JS package. This is currently implemented as one request/one user push.

There are three (generalized) types of push notifications that we send/would like to send:

1) Someone liked your post – basically a one-off push to a specific user based on a specific action. This we currently have implemented via Postgres trigger which uses `net.http_post` to send the request via GCF / FCM

2) New object inserted, notify followers – for some activities, we want to notify all of a user’s followers that they have taken some action. We started implementing this with a FOR LOOP in the Postgres trigger to call `net.http_post` for every follower. Our challenge here was that for users with a lot of followers, calling that many requests would cause the trigger to timeout, which rolled back the entire transaction. This was obviously problematic (especially if they retried the action, and re-pushed the first ~100 users again!!)

3) Send notification to lots of people/whole user base – occasionally, for various promotions or special holidays, etc., we would want to send a push notification to our entire user base (or some segment). To that end, we’ve started implementing a basic admin panel, but have not fully implemented yet. It’s a bit of a heavy lift, so just wanted to think through push holistically before completing that work.

My biggest concern is that having the triggers responsible for push is problematic if the trigger fails for any reason (either because of timeout, or just our GCF not responding) and then rolls back the transaction. I would prefer to decouple that logic – clearly for (2) in bulk cases, but also probably for (1) to be safer.

Some ideas we’ve explored:
- implementing an explicit `push_notifications` table with the necessary data, and then a realtime listener that queries for new entries to that table
- “batching” our pushes and passing the “batch” data to our GCF instead (might run into request size trouble as we get into larger batches if we’re not thoughtful here)

Mostly surfacing this as a discussion and to see whether any other apps have a strategy that handles both realtime and batched push notifications more effectively!

**Comment by anngbaum** [comment]
was doing a little searching around discord to see if any others had tips here, and found [this thread](https://discord.com/channels/839993398554656828/1021927855220871188/1021927855220871188) which echoed many of the same questions.

**Comment by caseycrogers** [comment]
I created a `notifications` table for handling push notifications. It has all the columns you need to send an FCM request to push to a user (`user_id`, `title`, `body`, `data`, etc).

Then it has a `state` and `retries` column. I have a trigger that runs a function that issues a rest request to FCM whenever a new row is inserted into this table. If the request fails, it's state gets set to `FAILED` and `retries` gets incremented. Then the same function also runs on a CRON to pick up failed rows and retry them so long as `retries` is below a certain threshold. If you really wanna get fancy with it you can add exponential back off here.

This system is pretty heavyweight, but it has the following advantages:
1. You can display a bell icon in the app with a list of recent notifications. You can even mark notifications as read/unread in the list.
2. You get retries, rate limiting, etc.
3. You have easier insight into what notifications are getting served.

You can also compute the `title` and `body` on the fly by putting all the data you want in the `data` column as a json blob (you can send `data` to FCM and it'll dutifully package it along with the notification). I used this to dynamically generate notification titles like "@<user_display_name> just tagged you!" (the user ID, not the display name, is stored in the json blob and then it's converted to a display name via a query).

This is especially useful if you have an in-app notification list where the notifications might be old and a username or other related content may have changed since the notification was originally triggered. And/or you decided to change the clientside copy for the notification. I actually have both the baked-to-string title and the underlying json blob so FCM can be lightweight and just directly fire off the raw string title and then the client notification list can do all the smart formatting. If it fails to properly parse the json because of a version change or something, it uses the string data as a fallback.

**Reply by anngbaum** [reply]
awesome write up, thank you @caseycrogers !

follow up question -- how many did you do at a time for your cron job?

**Reply by anngbaum** [reply]
asking because I assume the cron job has a maximum number of requests it can make before it times out, but I'm curious if you had a reliable way of staying within that range but still achieving the necessary throughput

**Reply by caseycrogers** [reply]
This is an old (now abandoned project) so I'm going back over it to see what I did. The above is just what I remember.

It looks like I killed the project halfway through migrating to the system I described earlier so I actually never got to adding the CRON and doing the retries/batching. I just had a trigger on the `notifications` table with one request per insert. Probably not scalable, thank god I wasn't getting traction ;)

**Reply by caseycrogers** [reply]
If you really need to battle harden everything you could shard your notifications table. I'm not sure where the bottlenecks are on Supa's end so that could mean sharding by notification ID and/or having sharded `notifications` tables or even having the notifications table and cron sharded across projects.

This is of course all pretty miserable so only recommended if you find you really need it to achieve scale-the Supa folks might have input here on where you're going to hit walls and whether or not such a nuclear approach is necessary.

**Comment by hosseinzare1** [comment]
Hi
Check out this article about setting up notifications from the Supabase docs:
[Sending Push Notifications](https://supabase.com/docs/guides/functions/examples/push-notifications?queryGroups=platform&platform=fcm)

Supabase recommends using a database webhook to trigger an Edge Function whenever a new item is added to the notifications table in order to send FCM notifications. You can set up a database webhook on the relevant table to automatically call an Edge Function and update your notifications table.

For example, if we want to notify users who have subscribed to a store whenever a new product is added to that store, we can create a database webhook on the Products table with an insert event. This webhook can call an Edge Function method that will identify the target users and update the notifications table accordingly.
